### PR TITLE
asyncfile: setLen procedure for files

### PR DESCRIPTION
**Pull request by ehmry**
Platform independent procedure to set a file length. Useful when
replacing file content.

### CODE REVIEW

File: tests/async/tasyncfile.nim [hunk 1]
```diff
@@ -24,7 +25,7 @@ proc main() {.async.} =
     var file = openAsync(fn, fmAppend)
     await file.write("\ntest2")
     let errorTest = file.readAll()
-    echo await errorTest
+    yield errorTest
```
Comment (dom96) [inline tests/async/tasyncfile.nim]: Why the change to yield here?

File: lib/pure/asyncfile.nim [hunk 2]
```diff
@@ -466,6 +466,15 @@ proc write*(f: AsyncFile, data: string): Future[void] =
       addWrite(f.fd, cb)
   return retFuture
 
+proc setLen*(f: AsyncFile, len: int64) =
```
Comment (dom96) [inline lib/pure/asyncfile.nim]: Does ``File`` support this operation as well?

File: tests/async/tasyncfile.nim (hunk 1 above)
Comment (ehmry) [inline tests/async/tasyncfile.nim]: The next line is ``doAssert errorTest.failed``, and running the test standalone results in an unhandled exception. Works both ways with ``koch test``.

File: lib/pure/asyncfile.nim (hunk 2 above)
Comment (ehmry) [inline lib/pure/asyncfile.nim]: Not that I can find, it seems you have to use something platform specific. I was planning on using ``AsyncFile`` for everything.

**Comment by ehmry** [comment]
Windows checks passed. I would add ``setLen`` for file, but I'm a bit confused about where I would do that. There is sysio.nim, but that seems POSIX specific.

I would be willing to move stuff like streams to use the multisync macro if I could understand and keep parity between the ``File`` and ``AsyncFile`` procs.

I'm partial to AsyncFile because I plan using a platform that does everything file-system in userspace, so the I/O is necessarily non-blocking in the applicaiton runtime to keep the everything running.

**Comment by ehmry** [comment]
@dom96 @Araq may I move the File type and procedures to a file.nim or filesystem.nim? Right now I think the docs are too spread out on this and I would like to be confident that the procedures are the same for both.

**Comment by dom96** [comment]
@ehmry If you can keep it backward compatible then I am very much happy for you to do that. Doing so will be challenging though, but could be done by exporting the procedures in system.nim. I have a feeling that @Araq will not like this though, so let's wait to see what he has to say.

**Comment by dom96** [comment]
I'm curious, what use case do you have for this ``setLen`` procedure?

**Comment by ehmry** [comment]
I want to completely replace the content of a file, so if the new content is shorter then I would have to delete and recreate the file or else truncate the file size to the size of the new content.

**Comment by ehmry** [comment]
May be better to simply make ``File | AsyncFile`` tests.

**Comment by dom96** [comment]
Wouldn't ``fmWrite`` achieve what you want?

**Comment by ehmry** [comment]
I'm not sure, I wouldn't want to rely on special POSIX open flags. The other use case is that I have a multiple buffers of known sizes that I will write in sequence, and I want to allocate space for the sum of the buffers so there is a greater chance the file will stay contiguous in storage.

I've also added a ``setLen`` to ``File``.

File: lib/pure/asyncfile.nim [hunk 3]
```diff
@@ -466,6 +466,38 @@ proc write*(f: AsyncFile, data: string): Future[void] =
...
+    low = getFileSize(f.fd.Handle, addr hi)
+    if low == INVALID_FILE_SIZE:
+      raiseOSError(osLastError())
+    result = (hi.int64 shr 32) or low.int64
+  elif defined(posix):
+    result = lseek(f.fd.cint, 0, SEEK_END)
+  elif defined(nimdoc):
+    discard
+
+proc setLen*(f: AsyncFile, length: Natural) =
```
Comment (Araq) [inline lib/pure/asyncfile.nim]: this should take an ``int64``.

File: lib/system/sysio.nim [hunk 4]
```diff
@@ -373,6 +375,11 @@ proc getFileSize(f: File): int64 =
   result = getFilePos(f)
   setFilePos(f, oldPos)
 
+proc setLen*(f: File, length: Natural) =
```
Comment (Araq) [inline lib/system/sysio.nim]: This should take an ``int64``.

**Comment by ehmry** [comment]
Length parameters are now int64 rather than Natural.

File: lib/pure/asyncfile.nim [hunk 5]
```diff
@@ -466,6 +466,38 @@ proc write*(f: AsyncFile, data: string): Future[void] =
       addWrite(f.fd, cb)
   return retFuture
 
+proc len*(f: AsyncFile): int64 =
```
Comment (dom96) [inline lib/pure/asyncfile.nim]: This procedure should also be added for the ``File`` type.

Comment (ehmry) [inline lib/pure/asyncfile.nim]: There is ``proc getFileSize*(f: File): int64 {.tags: [ReadIOEffect], benign.}`` in system and sysio, I could mark ``getFileSize`` as deprecated and define ``len`` for ``File``, or add ``getFileSize`` to ``AsyncFile``.

Comment (dom96) [inline lib/pure/asyncfile.nim]: ahh, I think we should have ``getFileSize`` and ``setFileSize`` for both. i.e. rename ``len`` to ``getFileSize`` and ``setLen`` to ``setFileSize``.

Comment (Araq) [inline lib/pure/asyncfile.nim]: Agreed. getFileSize and setFileSize are the names to use.

**Comment by ehmry** [comment]
Now ``getFileSize`` and ``setFileSize`` at ``File`` and ``AsyncFile``.

File: lib/pure/asyncfile.nim [hunk 6]
```diff
@@ -70,14 +70,18 @@ else:
...
-proc getFileSize(f: AsyncFile): int64 =
-  ## Retrieves the specified file's size.
-  when defined(windows) or defined(nimdoc):
-    var high: DWord
-    let low = getFileSize(f.fd.Handle, addr high)
+proc getFileSize*(f: AsyncFile): int64 =
+  ## Get a file Length.
+  when defined(windows):
+    var hi, low: int32
+    low = getFileSize(f.fd.Handle, addr hi)
```
Comment (dom96) [inline lib/pure/asyncfile.nim]: The changes you made here and above should be reverted.

File: lib/pure/asyncfile.nim [hunk 7]
```diff
@@ -70,14 +70,18 @@ else:
...
     if low == INVALID_FILE_SIZE:
       raiseOSError(osLastError())
-    return (high shl 32) or low
+    result = (hi.int64 shr 32) or low.int64
```
Comment (dom96) [inline lib/pure/asyncfile.nim]: Was the previous code that uses ``shl`` incorrect?

File: lib/pure/asyncfile.nim [hunk 8]
```diff
@@ -70,14 +70,18 @@ else:
...
+  elif defined(posix):
+    result = lseek(f.fd.cint, 0, SEEK_END)
+  elif defined(nimdoc):
```
Comment (dom96) [inline lib/pure/asyncfile.nim]: Remove this branch and change the above to an ``else``.

File: lib/pure/asyncfile.nim (hunk 7 above)
Comment (ehmry) [inline lib/pure/asyncfile.nim]: No, there was just a copy-and-paste.

File: lib/system/sysio.nim [hunk 9]
```diff
@@ -52,6 +52,8 @@ proc c_ferror(f: File): cint {.
   importc: "ferror", header: "<stdio.h>", tags: [].}
 proc c_setvbuf(f: File, buf: pointer, mode: cint, size: csize): cint {.
   importc: "setvbuf", header: "<stdio.h>", tags: [].}
+proc c_ftruncate(f: FileHandle, len: clong): cint {.
```
Comment (Araq) [inline lib/system/sysio.nim]: MS VC does not have <unistd.h>, so this PR causes Windows to break!

Comment (dom96) [inline lib/system/sysio.nim]: I didn't realise. Could we get App Veyor to test VC?

Comment (ehmry) [inline lib/system/sysio.nim]: I'm fine with reverting the sysio changes, so long as the platform specific ``AsyncFile`` stuff remains.

Comment (Araq) [inline lib/system/sysio.nim]: > Could we get App Veyor to test VC?

I'm investigating.

> I'm fine with reverting the sysio changes, so long as the platform specific AsyncFile stuff remains.

That is the case.

Comment (arnetheduck) [inline lib/system/sysio.nim]: ftruncate takes an `off_t` - what `off_t` is depends on platform and compile flags - like this, results may be mixed when setting file sizes >2gb as nim will silently truncate the value.

Comment (ehmry) [inline lib/system/sysio.nim]: Yes, and c_ftell should return clonglong as well.
